### PR TITLE
Adding Wayland support

### PR DESCRIPTION
**Pull request by 94824**
Hello again!

I was tinkering around, especially with the new gestures on Gnome 40--and really wanted this project to work on wayland. I wasn't able to figure how to do it natively, but it works very well under XWayland.

The problem was that self.screen.get_active_workspace() in Wnck didn't play nice, so I just used xprop. I believe that xprop should be available on any xorg setup, so I don't think a new dependency is required. 

Also, I'm not entirely sure if this setup works under hibernation, because I don't have it setup on my system--so it might break under that. 

Other than that, I just added some documentation to the existing VLC stuff, it was kinda bare and there was some stuff that didn't need to be there.

### CODE REVIEW

**Comment by jeffshee** [comment]
Hi, thanks again for another effort!
I definitely would like to support Wayland! Really hope that Nvidia can finally bring their cards to Wayland soon. ;-)
(It means that I don’t have a setup for Wayland development for now...)

If I understand you correctly, you mentioned that Hidamari actually works in Wayland session through XWayland, really? TBH I didn’t even think that it was possible, since other projects that draw on top of desktop (for example Komorebi and Conky) also having the same problem. If Hidamari can’t stick itself onto the root windows (desktop) in the first place, the support for Wayland would be meaningless.

For the maximized window detection, IIRC `libwnck` only works in X11. To implement such functionality in Wayland, I believe that we would need another solution, for example:
https://stackoverflow.com/questions/45465016/how-do-i-get-the-active-window-on-gnome-wayland

To summarize, the goal for Wayland support would be:
- [ ] Draw on top of desktop
- [ ] Get current active workspace
- [ ] Get the state for each window (maximized? Located in which workspace?)

**Comment by 94824** [comment]
Well, all of those goals are acheived--most of the work was already done (XWayland works wonders apparently)

Drawing on top of the desktop didn't require any new changes other than instructing the program to run under X11 under all situations, this forces the program to run under XWayland. That's done with `os.environ['GDK_BACKEND'] = "x11"` I was confused myself that it was able to draw on top without any additional tweaks--but I'll take not having to try and fix it.

Getting the active workspace is done via `xprop -root -notype _NET_CURRENT_DESKTOP`.

Getting the state for each window works under libwnck without modification. It seems that libwnck only had issues *getting* the workspace. Actually finding the maximized window works without modification once it's given a workspace via self.screen.get_workspace(x)

I can confirm on my systems that it works as intended, including the auto pausing on maximized--although it would be good to double check against another system, but both my PC and Laptop (Albiet both on Intel graphics so not much variability) work flawlessly.

**Comment by jeffshee** [comment]
I see. I will find a way to test it under Wayland, and hopefully we can get the workspace issue fixed. Just knowing that the basic works on Wayland already make me happy, what a time to be alive! 🥰

**Comment by jeffshee** [comment]
Hi, after some simple testing, I found some issues that we need to solve:

1. GPU decoding doesn't work when using Nvidia with Wayland. Actually there is nothing we can do now except waiting for the full Wayland support from Nvidia, hopefully in the next 470.x driver.
```
[00007f98f4005c40] gl gl: Initialized libplacebo v2.72.0 (API v72)
Xlib:  extension "NV-GLX" missing on display ":0".
Xlib:  extension "NV-GLX" missing on display ":0".
```

2. The current method of detecting maximized window actually works only if the applications run in XWayland. You can try with any application that support Wayland, it won't be detected unless forcing them to use XWayland with `GDK_BACKEND=x11`.

**Comment by 94824** [comment]
Leave it to Nvidia to break things :)

Thanks for finding that issue with xprop, In hindsight I should have realized that an X related tool wouldn't be able to properly detect wayland windows, guess I was just hoping XWayland did some magic behind the scenes.

I suppose we have two options in how we can proceed, We could either:

Figure out how to make Hidamari work natively under wayland, which is daunting to say the least. I'm truthfully not even sure if it's possible to run libvlc under a pure wayland environment, thanks to it hooking using get_xid and set_xwindow.

Or, we could have a "Partial Support" on Wayland. It wouldn't work on all hardware, such as Nvidia, and we would have to either disable maximized detection or just warn that it doesn't work correctly in all situations, but at least it would kinda work, and for some systems (such as mine) users could still use most of Hidamari without having to use X.

I think we can fix the window detection, I remember seeing a project on Github that needed to find the active window in a wayland environment and talked directly to mutter (https://github.com/ActivityWatch/aw-watcher-window/pull/46/commits/d3ee3c398a613fd7506c02fda9182101d3ad7eba), so maybe If we can figure out how to talk through the Dbus interface we can grab windows and see if they're maximized?

**Comment by 94824** [comment]
Little progress update.  I dug through GJS to try and talk to the shell, and I managed to get it to work. Unfortunately, WindowHandler still uses Wnck, which means that the check will still only occur when an X window is changed. However, the check itself will find wayland windows.

So, if you have a max wayland window and move a X window, Hidamari will find the wayland window and pause playback. However, it won't pause playback until a X window is created/destroyed/moved because the Wnck hooks on the WindowManager only work on X. 

So, obviously the next step is to migrate away from Wnck to something that can notice when wayland windows alongside x windows are modified--but I'm not sure what we should migrate to.

I was thinking that maybe we could have the check() function on the WindowManager be separate and simply run in the background of the application every second or so, but I'm not sure how to accomplish that. I don't think having it run like that would incur much performance loss, and it everything should work without wnck even being required.

**Comment by jeffshee** [comment]
Hi,
With the above proposal, I implemented the `WindowHandler` that specially made for Gnome.
I haven't integrated the code to Hidamari yet, but it seems to work fine alone already.

Here is the code snippet:
```python
import gi
import pydbus

gi.require_version('Gtk', '3.0')
from gi.repository import GLib, Gtk


class WindowHandlerGnome:
    def __init__(self, on_window_state_changed: callable):
        self.on_window_state_changed = on_window_state_changed
        self.gnome_shell = pydbus.SessionBus().get("org.gnome.Shell")
        self.prev_state = None
        GLib.timeout_add(500, self.eval)

    def eval(self):
        is_changed = False

        # Previous method using xprop (possible instability under Wayland?)
        # workspace = int(subprocess.getoutput("xprop -root -notype _NET_CURRENT_DESKTOP").split(" = ")[-1])

        ret1, workspace = self.gnome_shell.Eval("""
                        global.workspace_manager.get_active_workspace_index()
                        """)

        ret2, maximized = self.gnome_shell.Eval("""
                var window_list = global.get_window_actors().find(window =>
                    window.meta_window.maximized_horizontally &
                    window.meta_window.maximized_vertically &
                    window.meta_window.get_workspace().workspace_index == {0}
                );
                window_list
                """.format(workspace))

        ret3, fullscreen = self.gnome_shell.Eval("""
                var window_list = global.get_window_actors().find(window =>
                    window.meta_window.is_fullscreen() &
                    window.meta_window.get_workspace().workspace_index == {0}
                );
                window_list
                """.format(workspace))

        if not all([ret1, ret2, ret3]):
            raise RuntimeError("Cannot communicate with Gnome Shell!")

        cur_state = {'is_any_maximized': maximized != "", 'is_any_fullscreen': fullscreen != ""}
        if self.prev_state is None or self.prev_state != cur_state:
            is_changed = True
            self.prev_state = cur_state

        if is_changed:
            self.on_window_state_changed({'is_any_maximized': maximized != "", 'is_any_fullscreen': fullscreen != ""})

        return True


if __name__ == "__main__":
    handler = WindowHandlerGnome(print)
    Gtk.main()
```

Also, I'm thinking to preserve the previous method using `Wnck` as a backup as well. Just in case if there is anyone using Hidamari with another DE.

**Comment by 94824** [comment]
Phenomenal job! I didn't know there was a way to query the active workspace using the Dbus interface, so I'm glad we can avoid using Xprop. I tested your code in Hidamari and it works perfectly to what I can see--including pausing Wayland windows.

I think keeping Wnck would be a good idea, although Gnome would be the only one that would support wayland to this extend. We could use:

```
if os.environ["DESKTOP_SESSION"] == "gnome":
    self.window_handler = WindowHandlerGnome(self._on_window_state_changed)
else:
    self.window_handler = WindowHandler(self._on_window_state_changed)
``` 

In the player, or something to that extent. Do you think it would be better to outright bar wayland use on non-gnome environments, or simply offer limited support. Maybe simply disabling the max window detection under wayland?

**Comment by 94824** [comment]
Also, I believe we could also support Cinnamon because it's derived from Gnome and has a simliar Dbus interface using `org.Cinnamon`. (https://blog.fpmurphy.com/2012/02/d-bus-cinnamon-and-the-gnome-shell.html) However, I don't have a computer running Cinnamon, so I can't test it.

**Comment by jeffshee** [comment]
For now, I think disabling the maximize window detection on non-gnome Wayland environment would be a better option. Just did a quick search and couldn’t find any way to achieve the same thing on KDE Wayland. For other DE that doesn’t have a feasible solution yet, it couldn’t be helped. They would need to wait for such API get implemented on their side. ;)
